From 1fa400c8c3384c61e8709f68ef75aac09fe27d6c Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Wed, 19 Apr 2023 15:59:52 +0800
Subject: [PATCH 1628/1646] [Huawei] mm/dpool: Support hwpoison for dynamic
 hugetlb pages

Offering: HULK
hulk inclusion
category: feature
bugzilla: 188703

--------------------------------

Current dhugetlb doesn't support hwpoison, which leads to two issues:

1. Memory failure cannot handle poison in free dynamic hugetlb freelist.
2. Once UCE occurs on a page belonging to dynamic hugetlb, the
   corresponding memcg cannot be deleted.

In order to solve this, for free dynamic hugetlb page, it is safe since
check_new_page() during page allocation can be the gatekeeper. During the
removing of memcg, it will drop the poisoned pages to ensure that this will
not the removing.

Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 include/linux/dynamic_hugetlb.h |  1 +
 include/linux/mm.h              |  1 +
 include/ras/ras_event.h         |  1 +
 mm/dynamic_hugetlb.c            |  8 +++++++-
 mm/memory-failure.c             | 11 +++++++++++
 5 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/include/linux/dynamic_hugetlb.h b/include/linux/dynamic_hugetlb.h
index 95d5878d4dfa..b2ee43536308 100644
--- a/include/linux/dynamic_hugetlb.h
+++ b/include/linux/dynamic_hugetlb.h
@@ -96,6 +96,7 @@ struct dhugetlb_pool {
 	/* The following kabi change is used for dynamic pool */
 	KABI_EXTEND(int range_cnt)
 	KABI_EXTEND(struct range *pfn_ranges)
+	KABI_EXTEND(long nr_poisoned_pages)
 
 	/*
 	 * The percpu_pool[] should only be used by dynamic hugetlb core.
diff --git a/include/linux/mm.h b/include/linux/mm.h
index 080bc8b82aba..ca516f7c542e 100644
--- a/include/linux/mm.h
+++ b/include/linux/mm.h
@@ -3190,6 +3190,7 @@ enum mf_action_page_type {
 	MF_MSG_BUDDY_2ND,
 	MF_MSG_DAX,
 	MF_MSG_UNSPLIT_THP,
+	MF_MSG_FREE_DPOOL,
 	MF_MSG_UNKNOWN,
 };
 
diff --git a/include/ras/ras_event.h b/include/ras/ras_event.h
index d8890dcc0e7c..8cff165a3aa3 100644
--- a/include/ras/ras_event.h
+++ b/include/ras/ras_event.h
@@ -414,6 +414,7 @@ TRACE_EVENT(aer_event,
 	EM ( MF_MSG_BUDDY_2ND, "free buddy page (2nd try)" )		\
 	EM ( MF_MSG_DAX, "dax page" )					\
 	EM ( MF_MSG_UNSPLIT_THP, "unsplit thp" )			\
+	EM ( MF_MSG_FREE_DPOOL, "free dynamic hugetlb page" )		\
 	EMe ( MF_MSG_UNKNOWN, "unknown page" )
 
 /*
diff --git a/mm/dynamic_hugetlb.c b/mm/dynamic_hugetlb.c
index 660bf2f1cbe3..56c4e8c1d196 100644
--- a/mm/dynamic_hugetlb.c
+++ b/mm/dynamic_hugetlb.c
@@ -1290,6 +1290,7 @@ static int migrate_used_pages(struct dhugetlb_pool *hpool)
 
 	lockdep_assert_held(&hpool->lock);
 
+	hpool->nr_poisoned_pages = 0;
 	for (i = 0; i < range_cnt; i++) {
 		struct range *range = &hpool->pfn_ranges[i];
 		u64 pfn;
@@ -1297,6 +1298,11 @@ static int migrate_used_pages(struct dhugetlb_pool *hpool)
 		for (pfn = range->start; pfn <= range->end; pfn++) {
 			struct page *page = pfn_to_page(pfn);
 
+			if (PageHWPoison(page)) {
+				hpool->nr_poisoned_pages++;
+				continue;
+			}
+
 			/* Unlock and try migration. */
 			spin_unlock(&hpool->lock);
 			if (page->mapping != NULL)
@@ -1343,7 +1349,7 @@ static int test_percpu_pool_empty(struct dhugetlb_pool *hpool)
 	/* lock again */
 	spin_lock(&hugetlb_lock);
 
-	return nr_used_pages == 0 ? 0 : -ENOMEM;
+	return (nr_used_pages - hpool->nr_poisoned_pages) == 0 ? 0 : -ENOMEM;
 }
 
 struct dhugetlb_pool_ops list_pool_ops = {
diff --git a/mm/memory-failure.c b/mm/memory-failure.c
index cd09c4ab6ba0..9cd3a90e5c67 100644
--- a/mm/memory-failure.c
+++ b/mm/memory-failure.c
@@ -745,6 +745,7 @@ static const char * const action_page_types[] = {
 	[MF_MSG_BUDDY_2ND]		= "free buddy page (2nd try)",
 	[MF_MSG_DAX]			= "dax page",
 	[MF_MSG_UNSPLIT_THP]		= "unsplit thp",
+	[MF_MSG_FREE_DPOOL]		= "free dynamic hugetlb page",
 	[MF_MSG_UNKNOWN]		= "unknown page",
 };
 
@@ -1641,6 +1642,8 @@ int memory_failure(unsigned long pfn, int flags)
 	 *    Implies some kernel user: cannot stop them from
 	 *    R/W the page; let's pray that the page has been
 	 *    used and will be freed some time later.
+	 * 3) it's a free page in hugetlb pool, and therefore in safe hand:
+	 *    check_new_page() during page allocation will be gate keeper.
 	 * In fact it's dangerous to directly bump up page count from 0,
 	 * that may make page_ref_freeze()/page_ref_unfreeze() mismatch.
 	 */
@@ -1661,6 +1664,14 @@ int memory_failure(unsigned long pfn, int flags)
 			}
 			action_result(pfn, MF_MSG_BUDDY, res);
 			res = res == MF_RECOVERED ? 0 : -EBUSY;
+		} else if (page_belong_to_dynamic_hugetlb(p)) {
+			/* pages belongs to dhugetlb can be dropped safely */
+			if (!PagePool(p)) {
+				action_result(pfn, MF_MSG_FREE_DPOOL,
+					      MF_RECOVERED);
+				res = 0;
+				goto unlock_mutex;
+			}
 		} else {
 			action_result(pfn, MF_MSG_KERNEL_HIGH_ORDER, MF_IGNORED);
 			res = -EBUSY;
-- 
2.17.1

